### PR TITLE
SISRP-41873 Replaces parts of FA API with SISEDO

### PR DESCRIPTION
**Pull request by marcpechaitis**
https://jira.berkeley.edu/browse/SISRP-41873

### CODE REVIEW

File: app/models/financial_aid/my_terms_and_conditions.rb [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+    def get_feed_internal
+      return {} unless is_feature_enabled
+      {
+        termsAndConditions: terms_and_conditions
+      }
+    end
+
+    def instance_key
+      "#{@uid}-#{my_aid_year}"
```
Comment (redconfetti) [inline app/models/financial_aid/my_terms_and_conditions.rb]: @marcpechaitis - The Terms and Conditions are related to a specific Aid Year? I thought it applied to all of them across the board?

Comment (marcpechaitis) [reply app/models/financial_aid/my_terms_and_conditions.rb]: @redconfetti yes Terms & Conditions are aid year specific.  You might be getting mixed up with Title IV, which spans every aid year?

File: app/models/edo_oracle/financial_aid/queries.rb [hunk 2]
```diff
@@ -191,6 +191,59 @@ def self.get_financial_aid_summary(person_id, aid_year)
         SQL
         result.first
       end
+
+      def self.get_aid_years(person_id)
+        safe_query <<-SQL
+        SELECT UC.CAMPUS_ID,
+          UC.STUDENT_ID,
+          UC.INSTITUTION,
```
Comment (lyttam) [inline app/models/edo_oracle/financial_aid/queries.rb]: @marcpechaitis You can remove INSTITUTION and any other columns that aren't actually being used from this SELECT - I think it's a best practice even if they are not being included in the response that gets sent to the UI.

File: spec/ui_selenium/my_finances_finaid_summary_spec.rb [hunk 3]
```diff
@@ -279,11 +279,11 @@
                       it ("offers a 'Details' link to UID #{uid}") { expect(has_details_link).to be true }
 
                       details_url = page.finaid_details_link_element.attribute('href')
-                      it ("offers a 'Details' link pointing to the right aid year to UID #{uid}") { expect(details_url).to eql("#{WebDriverUtils.base_url}/finances/finaid/#{year_id}") }
+                      it ("offers a 'Details' link pointing to the right aid year to UID #{uid}") { expect(details_url).to eql("#{WebDriverUtils.base_url}/finances/financial_aid/#{year_id}") }
 
                       if api_shopping_sheet_url.nil?
                         awards_url = page.awards_link_element.attribute('href')
-                        it ("offers an 'Awards' link pointing to the right aid year to UID #{uid}") { expect(awards_url).to eql("#{WebDriverUtils.base_url}/finances/finaid/#{year_id}") }
+                        it ("offers an 'Awards' link pointing to the right aid year to UID #{uid}") { expect(awards_url).to eql("#{WebDriverUtils.base_url}/finances/financial_aid/#{year_id}") }
```
Comment (lyttam) [inline spec/ui_selenium/my_finances_finaid_summary_spec.rb]: @marcpechaitis this needs to stay `finaid` - and on line 282 also. It's referencing a URL to a page.

Also, these files in `spec/ui_selenium` are Selenium tests that we've let go to seed. We don't run them anymore, they are mostly here for historical reasons I think (and we should probably remove them, since they can be found in GitHub).

File: spec/ui_selenium/pages/my_finances_financial_aid_page.rb [hunk 4]
```diff
@@ -194,7 +194,7 @@ def coa_grand_total
 
       # PROFILE CARD
 
-      div(:profile_msg, :class => 'cc-widget-finaid-profile-message')
+      div(:profile_msg, :class => 'cc-widget-financial_aid-profile-message')
```
Comment (lyttam) [inline spec/ui_selenium/pages/my_finances_financial_aid_page.rb]: @marcpechaitis these also need to be changed back

File: app/models/edo_oracle/financial_aid/queries.rb (hunk 2 above)
Comment (marcpechaitis) [reply app/models/edo_oracle/financial_aid/queries.rb]: Thanks @lyttam, I have removed these references

File: src/assets/javascripts/angular/controllers/pages/termsAndConditionsController.js [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        $scope.termsAndConditions.isLoading = false;
+      }
+    );
+  };
+
+  getTermsAndConditions();
+
+  $scope.$on('calcentral.custom.api.finaid.approvals', function() {
+    termsAndConditionsFactory.getTermsAndConditions({
+      refreshCache: true
```
Comment (lyttam) [inline src/assets/javascripts/angular/controllers/pages/termsAndConditionsController.js]: @marcpechaitis you can remove the `refreshCache` parameter since you're using the ENF to do that.

File: src/assets/javascripts/angular/controllers/widgets/finaid/finaidApprovalsController.js [hunk 6]
```diff
@@ -26,17 +26,8 @@ angular.module('calcentral.controllers').controller('FinaidApprovalsController',
...
-        finaidFactory.getSummary({
-          refreshCache: true
-        });
-        showDeclineMessage(data);
-      } else {
-        sendEvent();
-        $location.path('/finances');
-      }
+    finaidFactory.postT4Response(response).then(function() {
+      sendEvent();
```
Comment (lyttam) [inline src/assets/javascripts/angular/controllers/widgets/finaid/finaidApprovalsController.js]: @marcpechaitis you could condense this to
```finaidFactory.postT4Response(response).then(sendEvent);```

File: src/assets/templates/finaid_t4.html [hunk 7]
```diff
@@ -7,30 +7,14 @@ <h1 class="cc-heading-page-title cc-print-hide">
...
-          data-button-text-approve="'Authorize'"
-          data-button-action-dont-approve="sendResponseT4('N')"
-          data-button-action-approve="sendResponseT4('Y')"
-          data-button-go-back="true"
-          data-can-post="api.user.profile.actAsOptions.canPost && !finaidSummary.title4.approved"
-        ></div>
+      <div class="medium-8 medium-centered columns cc-column-3 cc-clearfix-container">
+        <div class="cc-widget">
+          <div class="cc-widget-title">
+          <!--<h2 data-ng-bind="currentCategory.name"></h2>-->
```
Comment (lyttam) [inline src/assets/templates/finaid_t4.html]: @marcpechaitis Don't forget to remove this commented markup.

File: src/assets/templates/widgets/profile/title4.html [hunk 8]
```diff
@@ -3,68 +3,80 @@
...
-      </span>
-
-      <button class="cc-button-link" data-ng-if="!title4.showMessage" data-ng-click="title4.showMessage = !title4.showMessage">Show more</button>
-      <button class="cc-button-link" data-ng-if="title4.showMessage" data-ng-click="title4.showMessage = !title4.showMessage">Show less</button>
-    </p>
-
+    <div
+      class="cc-text-pre-line"
+      data-ng-bind="title4.longTitle"
+      data-ng-if="title4.longTitle"
```
Comment (lyttam) [inline src/assets/templates/widgets/profile/title4.html]: @marcpechaitis I don't think you need the `data-ng-if` attribute - if `title.longTitle` evaluates to false, then angular will render an empty `<div>` right?

**Comment by lyttam** [comment]
@marcpechaitis Finished reviewing the UI changes. Thanks for all your work! This will be a huge improvement to the gnarly Financial Aid code.

File: app/models/financial_aid/my_aid_years.rb [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+    def default_aid_year
+      return nil unless is_feature_enabled
+
+      feed = self.get_feed
+
+      default_aid_year = feed[:aidYears].select do |year|
+        !!year[:defaultAidYear]
+      end
+      default_aid_year[0][:aidYear]
```
[the remark this answers is not in the document]
Comment (marcpechaitis) [reply app/models/financial_aid/my_aid_years.rb]: Thanks for the detailed explanation @lyttam, it turns out we need to use `default_aid_year[0].try(:[], :id)` instead.

Hat tip to @davidlimcheng

File: app/models/financial_aid/my_aid_years.rb [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+        aid_received_spring_boolean = ActiveRecord::Type::Boolean.new.type_cast_from_database(year['aid_received_spring'])
+        aid_received_summer_boolean = ActiveRecord::Type::Boolean.new.type_cast_from_database(year['aid_received_summer'])
+
+        aid_year[:availableSemesters] << 'Fall' if aid_received_fall_boolean
+        aid_year[:availableSemesters] << 'Spring' if aid_received_spring_boolean
+        aid_year[:availableSemesters] << 'Summer' if aid_received_summer_boolean
+
+        result << aid_year
+
+      end
```
[the remark this answers is not in the document]
Comment (marcpechaitis) [reply app/models/financial_aid/my_aid_years.rb]: @lyttam, I have taken your suggestion and come up with this
```
    def aid_years
      result = EdoOracle::FinancialAid::Queries.get_aid_years(@uid).map do |year|
        aid_year = {
          id: year['aid_year'],
          name: year['aid_year_descr'],
          defaultAidYear: ActiveRecord::Type::Boolean.new.type_cast_from_database(year['default_aid_year']),
          availableSemesters: []
        }

        #convert string to boolean
        aid_year[:availableSemesters] << 'Fall' if ActiveRecord::Type::Boolean.new.type_cast_from_database(year['aid_received_fall'])
        aid_year[:availableSemesters] << 'Spring' if ActiveRecord::Type::Boolean.new.type_cast_from_database(year['aid_received_spring'])
        aid_year[:availableSemesters] << 'Summer' if ActiveRecord::Type::Boolean.new.type_cast_from_database(year['aid_received_summer'])
        aid_year
      end
    end
```

File: app/models/financial_aid/my_title4.rb [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+      {
+        title4: title4
+      }
+    end
+
+    private
+
+    def title4
+      title4 = EdoOracle::FinancialAid::Queries.get_title4(@uid)
+      return nil unless title4
```
[the remark this answers is not in the document]
Comment (marcpechaitis) [reply app/models/financial_aid/my_title4.rb]: @lyttam, thanks again... I have made this change

File: spec/models/edo_oracle/financial_aid/queries_spec.rb [hunk 12]
```diff
@@ -128,4 +128,53 @@
...
+
+    it 'returns the expected result' do
+      expect(subject.count).to eq 7
+      expect(subject).to have_keys(%w(aid_year approved response_descr main_header main_body dynamic_header dynamic_body))
+    end
+
+    # it 'sorts the rows by aid year' do
+    #   expect(subject[0]['aid_year']).to eq '2018'
+    #   expect(subject[1]['aid_year']).to eq '2019'
+    # end
```
[the remark this answers is not in the document]
Comment (marcpechaitis) [reply spec/models/edo_oracle/financial_aid/queries_spec.rb]: @lyttam, thanks again... I have made this change

File: app/models/financial_aid/my_aid_years.rb (hunk 10 above)
[the remark this answers is not in the document]
Comment (lyttam) [reply app/models/financial_aid/my_aid_years.rb]: @marcpechaitis Looks good, the only other thing I would suggest is refactoring so that you don't need that comment. Instead of a comment saying what the code does, put the code in a function with a descriptive name:

```
aid_year[:availableSemesters] << 'Fall' if convert_string_to_boolean(year['aid_received_fall'])
```
```
def convert_string_to_boolean(str)
  ActiveRecord::Type::Boolean.new.type_cast_from_database(str)
end
```

Also you might want to write a unit test to find out what happens if `year['aid_received_fall']` is `nil` - will `#type_cast_from_database` throw an exception? Or will it evaluate `nil` as `false`?

File: spec/ui_selenium/my_finances_finaid_summary_spec.rb (hunk 3 above)
Comment (marcpechaitis) [reply spec/ui_selenium/my_finances_finaid_summary_spec.rb]: @lyttam, thanks again... I have made this change

File: spec/ui_selenium/pages/my_finances_financial_aid_page.rb (hunk 4 above)
Comment (marcpechaitis) [reply spec/ui_selenium/pages/my_finances_financial_aid_page.rb]: @lyttam, thanks again... I have made this change

File: src/assets/javascripts/angular/controllers/pages/termsAndConditionsController.js (hunk 5 above)
Comment (marcpechaitis) [reply src/assets/javascripts/angular/controllers/pages/termsAndConditionsController.js]: @lyttam, thanks again... I have made this change

File: src/assets/javascripts/angular/controllers/widgets/finaid/finaidApprovalsController.js (hunk 6 above)
Comment (marcpechaitis) [reply src/assets/javascripts/angular/controllers/widgets/finaid/finaidApprovalsController.js]: @lyttam, thanks again... I have made this change

File: src/assets/templates/finaid_t4.html (hunk 7 above)
Comment (marcpechaitis) [reply src/assets/templates/finaid_t4.html]: @lyttam, thanks again... I have made this change

File: src/assets/templates/widgets/profile/title4.html (hunk 8 above)
Comment (marcpechaitis) [reply src/assets/templates/widgets/profile/title4.html]: @lyttam, thanks again... I have made this change

File: src/assets/templates/directives/finaid_permissions.html [hunk 13]
```diff
@@ -7,22 +7,26 @@ <h2 data-ng-bind="title"></h2>
...
-        <button
-          class="cc-button"
-          data-ng-bind="buttonTextDontApprove"
-          data-ng-click="buttonActionDontApprove()"
-          data-ng-disabled="!canPost"
-          data-ng-if="buttonTextDontApprove"
-        ></button>
+      <h3 data-ng-bind="responseHeader" data-ng-if="responseHeader"></h3>
+      <div class="cc-text-pre-line" data-ng-bind="responseText" data-ng-if="responseText"></div>
+      <p>
```
[the remark this answers is not in the document]
Comment (marcpechaitis) [reply src/assets/templates/directives/finaid_permissions.html]: @lyttam, thanks again... I have made this change

File: src/assets/javascripts/angular/factories/finaidFactory.js [hunk 14]
```diff
@@ -17,7 +17,7 @@ angular.module('calcentral.factories').factory('finaidFactory', function(apiServ
   // var urlFinaidYear = '/dummy/json/financial_aid_data.json';
   // var urlFinaidYear = '/dummy/json/financial_aid_data_summer_only.json';
   var urlFinancialAidSummary = 'api/my/financial_aid_summary';
-  var urlSummary = '/api/campus_solutions/aid_years';
```
[the remark this answers is not in the document]
Comment (marcpechaitis) [reply src/assets/javascripts/angular/factories/finaidFactory.js]: @redconfetti Yes the endpoint you reference is still necessary, because we have replaced only a small part of the "FA API" with the changes I introduce.  Future sprints will accomplish the replacement of the rest of it.

For now we need to leave it in.  Right, @lyttam ?

**Comment by redconfetti** [comment]
@marcpechaitis - Please focus on making the updates noted in a follow-up PR. I'm going to merge this now just so that the BSA can begin testing sooner

**Comment by redconfetti** [comment]
@marcpechaitis 

I think you should be able to refactor CampusSolutions::TermsAndConditions (the posting proxy) to use FinancialAid::MyAidYears (SISEDO aid years model), and then you can get rid of:

* This route `/api/campus_solutions/aid_years` in routes.rb
* CampusSolutions::AidYearsController (used entirely for that route)
* CampusSolutions::MyAidYears (cacheing model for old CS API Aid Years)
* CampusSolutions::AidYears (proxy model for old CS API aid years)

Then I would also create a JIRA to have PS Dev team investigate who consumes `UC_FA_GET_T_C.v1`, and if it's only CalCentral you can deprecate it completely.

Also please address @lyttam 's [comments above](https://github.com/ets-berkeley-edu/calcentral/pull/7395#discussion_r233599661) in your follow-up PR.
